README.md
# emacs.d_wdp
it's my current can normally work emacs.d files. it based on chenbin's emacs.d in 2014. it' s old, but adapt me.
the chinbin's lastest emacs.d is good, but it's slow for me. and some operation habit is different.


history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(buffer-name-history tmm--history query-replace-history query-replace-from-to-history face-name-history regexp-history woman-topic-history search-ring coding-system-history regexp-search-ring extended-command-history dired-shell-command-history string-rectangle-history org-read-date-history tmphist bmkp-tag-history evil-ex-history ido-buffer-history minibuffer-history file-name-history ido-file-history))
(setq buffer-name-history '("lpbk_y1564_det_transfer_function.sv • eth-fpga3" "def.svh" "relative_config_constraints.sv" "rif_proc_driver_item_1.sv" "relative_config.sv" "port_list.svh" "tb.svh" "vlan_det_lookup_provisioning.sv" "vlan_det_lookup_tb_def.svh" "vliw_class_extends.sv" "tc_008.log" "mac_tx_3.sv" "vopt.log" "pipe_gen.rb" "pipe_gen_circuit_spec.rb" "patch_def_svh.rb" "metadata_gen.rb" "makefile • scripts" "database.mk • B_xgPro_v15.0" "xgPro-files.mk" "tdm_wrapper-files.mk" "revision-cntl.mk" "device_reg-files.mk" "vlog_macro_defined.f • tc_007" "tc_007.log" "top.sv • sim_uvm" "tc_004.log"))
(setq tmm--history '("h==>Help" "y==>YASnippet" "v==>Verilog" "s==>Statements" "t==>Tools" "b==>Buffers" "o==>Options" "e==>Edit" "f==>File" "h==>Help" "y==>YASnippet" "v==>Verilog" "s==>Statements" "t==>Tools" "b==>Buffers" "o==>Options" "e==>Edit" "f==>File" "h==>Help" "y==>YASnippet" "v==>Verilog" "s==>Statements" "t==>Tools" "b==>Buffers" "o==>Options" "e==>Edit" "f==>File" "h==>Help" "y==>YASnippet" "v==>Verilog" "s==>Statements" "t==>Tools" "b==>Buffers" "o==>Options" "e==>Edit" "f==>File"))
(setq query-replace-history '("lastSegSize" "frmSize" "readBackPutProcItem()" "putProcItem()" "injectionFrame" "dyingGaspFrame" "calculatedBandwidthPerPortInterval" "calculatedBandwidthPerPort" "calculatedBandwidthInterval" "calculatedBandwidth" "calculatedBandwidthPerPortAverage" "calculatedBandwidthPerPortTemp1" "calculatedBandwidthAverage" "calculatedBandwidthTemp1" "[37:0]" "[34:0]" "rate_meter_multiple" "PTP_PROCESS" "rate_meter_multiple_transfer_function" "ptp_process_transfer_function" "mac_rx_" "Rate Meter Multiple" "Rate Meter RX " "RateMeterMultiple" "RateMeterRx" "rate_meter_rx" "bypassCurrentPacket =1;" "return;" "inputItem" "InputItem" "correctionField" "igrPhyTs" "infoStringProv" "infoString" "PTP PROCESS" "PTP CNTL" "ptp_process" "ptp_cntl" "\"INGRESS\"" "\"UPPER\"" "INSTANCE_POSITION" "PARSER_POSITION" "Ptp Process " "Port Parser " "port_parser_transfer_function" "engineSel" "speedSel" "NUM_OF_CONTEXTS" "NUM_CONTEXTS" "NUMBER_OF_MACS" "cfg::NUM_OF_CONTEXTS" "a0bcde" "da0000" "48'hbacdef" "48'hda0000" "NUM_OF_CTX_FOR_ETH_FOR_TB_ENV" "NUM_OF_CTX_FOR_ETH" "transmitArbiter_u5" "transmitArbiter_u2" "receiveArbiter_u5" "receiveArbiter_u2" "l2Tuple.ethernet.packet" "l2Tuple.packet" "LPBK_Y1564_DET" "FAC_TERM_LPBK_DET" "LPBK_Y1564_CNTL" "FAC_TERM_LPBK_CNTL" "tfPrintMessageId" "printNameGetProcItem" "contextEna" "contextEnabled" "MAC_TX_" "MAC_RX_" "MAC_RX_CONTEXT_ATTRI" "MAC_RX_CONTEXT_ATTRI_VLAN_ETHERTYPE" "srcPort" "vliwSourcePortNum" "tfInternalIfTx" "tfSgmiiTx" "tfInternalIfRx" "tfSgmiiRx" "phyPortnum" "physicalRgmiiNumber" "INTERNAL TX " "INTERNAL GMII TX " "INTERNAL RX " "INTERNAL NGMII RX " "INTERNAL RX Transfer Function" "INTERNAL NGMII RX Transfer Function" "internal_tx_if_transfer_function" "internal_gmii_tx_if_transfer_function" "internal_rx_if_transfer_function" "internal_ngmii_rx_if_transfer_function" "`NUM_OF_PORTS" "`PORTS" "INTERNAL NGMII RX TF process access" "INTERNAL GMII RX TF process access" "INTERNAL GMII RX Transfer Function" "internal_gmii_rx_if_transfer_function" "TC_003" "TC_000" "$LOCAL_DB_PATH/$LOG_FILE" "./$LOG_FILE" "$LOCAL_DB_PATH/$LOG_FILE " "./$LOG_FILE " "$LOCAL_DB_PATH" "$CURRENT_PATH" "database_up" "database_update" "database_list" "all-source-svn-update" "cp-source-mk-list" "all-modules-svn-update" "cp-module-mk-list" "SVN_UPDATE" "LIST_FILE" "LOG_FILE" "SVN_UPDATE_LOG_FILE" "tmpBlockLockDuration" "tmpLockDuration" "tmpBlockLockBitPos" "tmpLockBitPos" "tmpBlockLocked" "tmpLocked" "blockLockContention" "lockContention" "blockLockDuration" "lockDuration" "blockLockBitPos" "lockBitPos" "blockLocked" "locked" "beforeAmLane" "pcsl" "stateRxFsm[pcsl]" "stateRxFsm" "txPmaBitPos" "pmaBitPos" "rxPmaBitPos" "caui4TxClkPos" "xxvTxClkPos" "caui4RxClkPos" "xxvRxClkPos" "caui4" "caui4sim" "xxv" "caui4sim_if" "xxv_if" "caui4TxFrame" "xxvTxFrame" "CGMII_" "XGMII_" "caui4Frame" "xxvFrame" "caui4IfDirection" "xxvIfDirection" "CAUI4_MON" "XXV_MON" "CAUI4 Monitor" "XXV Monitor" "Cgmii" "Xgmii" "cgmiiCode" "xgmiiCode" "cGBaseR" "hundredGBaseR" "tenGBaseR" "100GBASE-R " "10GBASE-R " "100GBaseR" "10GBaseR" "reserved100GBaseR" "reserved10GBaseR" "conv100GBaseRToCgmii" "conv10GBaseRToXgmii" "convCgmiiTo100GBaseR" "convXgmiiTo10GBaseR" "CGBASE" "XGBASE" "CGMII" "XGMII"))
(setq face-name-history '("hi-blue" "hi-green" "hi-pink" "hi-yellow" "hi-red-b" "hi-green-b" "hi-blue-b" "hi-black-b" "hi-black-hb"))
(setq regexp-history '("da a1 e2 e2" "e2 de c5 78 " "14 b5 f9 57" "b7 6a 72 60 f4" "3c 00 00 05" " ec 52 a1 d5 " "ec 52 a1 d5" "96 5e c9 29" "63 75 d6 77" "9d 16 d1 bc" "15 40 06 44" "73 55 1e 19" "d1 bc 9d 16" " 18 fc" "b5 68" "ab aa 51 95" "00 c1 2b c0" "60 4d 86 0b" "22 73 b1 48" "7f 96" "18 92" "ce ee d3 3b" "d3 3b ce ee" "60 1c 47 dd" "47 dd 60 1c" "bb c1 7a a2" "ec 30 5c c1" "6b 1b 3d 44" "64 b5 02 1f" "aa c3 ac c8" "1c 6f 4c 1d" "6b 1d 27 0b" "68 f5 72 5d" "46 47 48 49" " 70 c3 43 e4 " "c9 6a 01 66" "0a 0b" "6c fc" "74 f8" "f8 18 05 56" " 05 55" " de da" "65 d1 f4 21" "6e 08 4d 49" " 1b a8 e3 b0" "1b a8 e3 b0" "78 a6 05 d2" "d5 42 21 ca" "08 cb 10 0d" "47 ce 5f ce" " c7 6b 7b 37 " "70 c3 43 e4" "00 08 6a e5 00 6a 00 00" "5e 31 2d 23 34 e5 b4 f1 c9 8d 0a 7d b8 5e 05 32 ac b0 3b 28 dd 01" "8a 48 64 64 0e b4 " "8a 48 64 64 0e b4 

lisp/tab-display.el
;;; tab-display.el --- tab display code for managing tabs visually.

;; Copyright (C) 1997-2000 Free Software Foundation, Inc.

;; Author: Kevin A. Burton (burton@openprivacy.org)
;; Maintainer: Kevin A. Burton (burton@openprivacy.org)
;; Location: http://relativity.yi.org
;; Keywords: tab, width, display
;; Version: 1.3.2

;; This file is [not yet] part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify it under
;; the terms of the GNU General Public License as published by the Free Software
;; Foundation; either version 2 of the License, or any later version.
;;
;; This program is distributed in the hope that it will be useful, but WITHOUT
;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
;; FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
;; details.
;;
;; You should have received a copy of the GNU General Public License along with
;; this program; if not, write to the Free Software Foundation, Inc., 59 Temple
;; Place - Suite 330, Boston, MA 02111-1307, USA.

;;; Commentary:

;; This is an attempt to handle tab display in a way that makes things obvious
;; when working with buffers.  I personally don't like using hard \t chars in my
;; source files because of width differences.

;;; Install:
;;
;; You should turn font-lock-mode on globally in order to see your tabs in
;; color.

;;; History:
;;
;; Thu Oct 25 2001 11:53 PM (burton@openprivacy.org): changed the tab width to 8
;; which seems like this is more popular.
;;
;; Sun Jul 22 2001 03:43 AM (burton@relativity.yi.org): now keeping the menu
;; under tools.
;;
;; Sat Jan 13 18:14:31 2001 (burton): Version 1.3.1.  Fixed another small bug
;; with standard-display-table vs buffer-display-table.
;;
;; Wed Dec 27 12:48:27 2000 (burton): Version 1.3.0.  Was making a brand new
;; display table for every new buffer.  This was a bad idea because if the user
;; made any changes to standard-display-table tab-display killed this.  Now if
;; the buffer-display-table is null it gets the value of
;; standard-display-table.  This should have been the behavior all along.  For
;; example, some gnus e-mail needs escaping but tab-display will kill this.
;;
;; Mon Dec 25 13:38:18 2000 (burton): Created version 1.2.0.  Using
;; font-lock-mode as the hook used to turn on tab-display (since it requires its
;; use).  No messages when tab-display-mode is on.  -on and -off functions
;; renamed. 
;;
;; 1.1.0: Turned it (correctly) into a minor mode with a menu
;;
;; 1.0.0: init

;;; TODO:


(require 'easymenu)
(require 'font-lock)
(require 'disp-table)

;;; Code:
(defface tab-display-face '((t (:foreground "MediumBlue" :bold t)))
  "Face used to highlight the annotation lines to the left of the annotate buffer.")

(defvar tab-display-tab-width 8 "Tab width.  See `tab-width'.")
(make-variable-buffer-local 'tab-display-tab-width)

(defvar tab-display-mode-map (make-sparse-keymap) "Mode specific keymap for `section-mode'.")

(defvar tab-display-mode nil  "Non-nil means tab-display-mode local minor mode is enabled.")
(make-variable-buffer-local 'tab-display-mode)

(defvar tab-display-show nil "True if the user is currently showing their tabs.")
(make-variable-buffer-local 'tab-display-show)

(defvar tab-display-mode-string " Tabs" "String for the mode-line to indicate `tab-display-mode'.")

(defvar tab-display-mode-hook nil "Hook to run after tab-display is started.")

(defun tab-display-toggle-tab-width()
  "Toggle tab-width between four and eight chars"
  (interactive)
  
  (if (= tab-width 4)
      (setq tab-width 8)
    (setq tab-width 4))

  (if tab-display-show
      (tab-display-show))
 
  (message "tab-width is now %s" tab-width))

(defun tab-display-mode()
  (interactive)
  "Toggle using tab-display"

  (if tab-display-mode
      (tab-display-mode-off)
    (tab-display-mode-on)))

(defun tab-display-mode-on()
  "Turn ON tab-display"
  (interactive)

  (setq tab-display-mode 1)
  (tab-display-show)
  
  (run-hooks 'tab-display-mode-hook))

(defun tab-display-mode-off()
  "Turn OFF tab-display"
  (interactive)

  (setq tab-display-mode 0)
  (tab-display-hide))

(defun tab-display-toggle()
  ;;toggle hiding/showing tabs
  (interactive)
  
  (if tab-display-show
      (tab-display-hide)
    (tab-display-show)))

(defun tab-display-show()
  ;;show tabs

  (setq tab-display-show t)
  
  ;;set the buffer-display-table to use the width of `tab-width'"
  
  ;;setup buffer display options...
  (if (null buffer-display-table)
      (setq buffer-display-table (copy-sequence standard-display-table)))
  
  (aset buffer-display-table ?\t (vconcat (make-string tab-width ?>))))

(defun tab-display-toggle-indent-tabs-mode()
  "Enable/disable the use of tabs."
  (interactive)
  
  (if indent-tabs-mode
      (progn
        (setq indent-tabs-mode nil)
        (message "Tabs: off"))
    (progn 
      (setq indent-tabs-mode t)
      (message "Tabs: on"))))

(defun tab-display-hide()
  ;;hide tabs...

  (message "Tabs will be hidden.")
  (setq tab-display-show nil)
  (aset buffer-display-table ?\t nil))

(defun tab-display-mode-hook()
  "Highlight tabs in font-lock-mode turning on tabs mode if necessary."

  (if (or (and (not (string-match "\\*" (buffer-name (current-buffer))))
               (not (equal major-mode 'dired-mode)))
          (or (equal major-mode 'gnus-article-mode)
              (equal major-mode 'message-mode)))
      (progn 

        (tab-display-mode-on)
        
        ;; color the TAB character

        (font-lock-add-keywords nil '(("\\(\t\\\)" 1 'tab-display-face t)))
        (font-lock-fontify-buffer))))
  
;;add it as a minor mode.
(add-to-list 'minor-mode-alist
             '(tab-display-mode tab-display-mode-string))

(let ((a (assoc 'tab-display-mode minor-mode-map-alist)))
  (if a
      (setcdr a tab-display-mode-map)
    (add-to-list 'minor-mode-map-alist
                 (cons 'tab-display-mode tab-display-mode-map))))

;;add the tab highlight code to font-lock
(add-hook 'font-lock-mode-hook 'tab-display-mode-hook)

(easy-menu-add-item 'menu-bar-tools-menu
                    nil
                    '("Tab Display"
                      ["Hide/show tabs" tab-display-toggle t]
                      ["Toggle tab width (4/8)" tab-display-toggle-tab-width t]
                      ["Toggle hard tabs use: \\t" tab-display-toggle-indent-tabs-mode t]))

(provide 'tab-display)

;;; tab-display.el ends here
